src/test/java/org/jenkinsci/plugins/workflow/cps/DeserializationTest.java

```diff
@@ -0,0 +1,136 @@
+                p.setDefinition(new CpsFlowDefinition(LOAD_PREFIX + LOAD_SEMAPHORE + LOAD_SUFFIX, true));
+                b = p.scheduleBuild2(0).getStartCondition().get();
+                SemaphoreStep.waitForStart("wait/1", b);
+                story.j.assertLogContains("Pre-semaphore value is " + EXISTING_VAR_VALUE, b);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+                WorkflowRun b = p.getBuildByNumber(1);
+                SemaphoreStep.success("wait/1", null);
+                story.j.assertLogContains("a ran on value from b", story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b)));
+                story.j.assertLogContains("Post-semaphore value is " + EXISTING_VAR_VALUE, b);
+            }
+        });
+    }
+
+    @Test
+    public void clashingVariablesOnRestart() throws Exception {
+        String clashingValue = "0";
+        String sameVariable = EXISTING_VAR_NAME + " = " + clashingValue + ";";
+        // Testing with identical variable names injected via load step
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                story.j.jenkins.getWorkspaceFor(p).child(A_GROOVY).write(A_CONTENTS, null);
+                ScriptApproval.get().approveSignature("method groovy.lang.Binding getVariables");
+                story.j.jenkins.getWorkspaceFor(p).child(B_GROOVY).write(B_CONTENTS, null);
+                // Control case:
+                p.setDefinition(new CpsFlowDefinition(sameVariable + LOAD_PREFIX + LOAD_SUFFIX, true));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.assertLogContains("a ran on value from b", b);
+                story.j.assertLogContains("Pre-semaphore value is " + clashingValue, b);
+                // Test case:
+                p.setDefinition(new CpsFlowDefinition(sameVariable + LOAD_PREFIX + LOAD_SEMAPHORE + LOAD_SUFFIX, true));
+                b = p.scheduleBuild2(0).getStartCondition().get();
+                SemaphoreStep.waitForStart("wait/1", b);
+                story.j.assertLogContains("Pre-semaphore value is " + clashingValue, b);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+                WorkflowRun b = p.getBuildByNumber(1);
+                SemaphoreStep.success("wait/1", null);
+                story.j.assertLogContains("a ran on value from b", story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b)));
+                story.j.assertLogContains("Post-semaphore value is " + clashingValue, b);
```

Comment:
It's not obvious to me what is happening in this case. Did you check via a debugger that the value of `INJECTED_VAR` in `shell.getContext().getVariables()` was `PRE_EXISTING`  in `readResolve` before being replaced? I would've thought that it was already `0` before the restart.

Also, in regards to https://github.com/jenkinsci/workflow-cps-plugin/pull/348/files#r370153155, I think the loaded scripts would also need to echo `INJECTED_VAR` after the main script changes the value of `INJECTED_VAR` to show that the equality check passes and that the bindings were linked successfully.

Reply:
so i did use the debugger to check that the `INJECTED_VAR` was `PRE_EXISTING` before  the [`putAll`](https://github.com/car-roll/workflow-cps-plugin/blob/7af475fda8512d321dec623074f07b83211a03ba/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java#L180). 
I was testing the scenario where the load script changes an environment variables, and then that new value survives  a restart.  
So in this case,`INJECTED_VAR` is 0 before the restart, but then `GroovyShellDecorator` will set `INJECTED_VAR` to `PRE_EXISTING` at the restart again and then readresolve is supposed to reset `INJECTED_VAR` back to zero.

Reply:
> so i did use the debugger to check that the INJECTED_VAR was PRE_EXISTING before the putAll.

Ok, sounds good, thanks for clarifying.

As far as https://github.com/jenkinsci/workflow-cps-plugin/pull/348/files#r370153155, do any of the new tests fail if [these lines](https://github.com/jenkinsci/workflow-cps-plugin/blob/5db478ab5d6781696d9b61ce75fae57774871040/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java#L184-L185) are removed? If so, then I think we're good and injecting vars via `GroovyShellDecorator` works ok with the `load` step.

Reply:
Unfortunately, this test does not fail when that equality line is removed. I'm trying to see if i can get another test to properly handle that case.